lib/ABI/FunctionType-revng-c.cpp

```diff
@@ -0,0 +1,424 @@
+      auto &Current = Result.emplace_back();
+      if (Used.at(Register).Type().UnqualifiedType().get() != nullptr)
+        Current.Type() = Used.at(Register).Type();
+      else
+        Current.Type() = { defaultRegisterType(Register), {} };
+      Current.CustomName() = Used.at(Register).CustomName();
+    } else {
+      // This register is unused but we still have to create an argument
+      // for it. Otherwise the resulting function will not be compatible with
+      // the current one: if the "hole" is not preserved, there will be no way
+      // to recreate it at any point in the future, when, for example,
+      // converting back to the original-like representation.
+      auto &Current = Result.emplace_back();
+      if (Used.at(Register).Type().UnqualifiedType().get() != nullptr)
+        Current.Type() = Used.at(Register).Type();
+      else
+        Current.Type() = { defaultRegisterType(Register), {} };
+      Current.CustomName() = Used.at(Register).CustomName();
+    }
+  }
+
+  // Rewrite all the indices to make sure they are consistent.
+  for (auto Pair : llvm::enumerate(llvm::reverse(Result)))
+    Pair.value().Index() = Pair.index();
+
+  return Result;
+}
+
+std::optional<llvm::SmallVector<model::Argument, 8>>
+TDManager::tryConvertingStackArguments(model::QualifiedType StackArgumentTypes,
+                                       size_t IndexOffset) const {
+  // Qualifiers here are not allowed since this must point to a simple struct.
+  revng_assert(StackArgumentTypes.Qualifiers().empty());
+
+  auto *Unqualified = StackArgumentTypes.UnqualifiedType().get();
+  if (!Unqualified) {
+    // If there is no type, it means that the importer responsible for this
+    // function didn't detect any stack arguments and avoided creating
+    // a new empty type.
+    return llvm::SmallVector<model::Argument, 8>{};
+  }
+
+  auto *Pointer = llvm::dyn_cast<model::StructType>(Unqualified);
+  revng_assert(Pointer != nullptr,
+               "`RawFunctionType::StackArgumentsType` must be a struct");
+  const model::StructType &Types = *Pointer;
+
+  // If the struct is empty, it indicates that there are no stack arguments.
+  if (Types.Fields().empty()) {
+    revng_assert(Types.size() == 0);
+    return llvm::SmallVector<model::Argument, 8>{};
+  }
+
+  // Compute the alignment of the first argument.
+  uint64_t CurrentAlignment = *ABI.alignment(Types.Fields().begin()->Type());
+  if (!isZeroOrPowerOfTwo(CurrentAlignment)) {
+    revng_log(Log,
+              "The natural alignment of a type is not a power of two:\n"
+                << serializeToString(Types.Fields().begin()->Type()));
+    return std::nullopt;
+  }
+
+  // Process each of the fields (except for the very last one),
+  // converting them into arguments separately.
+  llvm::SmallVector<model::Argument, 8> Result;
+  for (auto Iterator = Types.Fields().begin();
+       Iterator != std::prev(Types.Fields().end());
+       ++Iterator) {
```

Comment:
It'd be nice to have `zip(skip(C), skipLast(C))`
See `include/revng/ADT/STLExtras.h:inline auto skip(unsigned ToSkip, C &&Container)`

Reply:
Here's how it turned out:
https://github.com/revng/revng/pull/232/commits/ecdcae6b135b0e6c4a4ac339f16efd51f7d7e83b#diff-869af126c25e96b2909972d4bfb0a8fbf2de2a1eb764a53975ddc300ee837cf5R309-R311
```cpp
// Look at all the fields pair-wise, converting them into arguments.
const auto &Range = llvm::zip(skip_back(Stack.Fields()),
                              skip_front(Stack.Fields()));
for (const auto &[CurrentArgument, NextArgument] : Range) {
```

but to make it happen I also had to extend the `skip` somewhat:
https://github.com/revng/revng/pull/232/commits/3701556c215ef26240d5cb9d8340bb7d48d81e25

What do you think?

Reply:
Since this seems to have passed the new review, I'm closing this thread.